gridfs.go

```diff
@@ -0,0 +1,391 @@
+	if b.opts.DryRun {
+		return &mockWriteCloser{}, nil
+	}
+
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem resolving bucket")
+	}
+
+	writer, err := grid.OpenUploadStream(name)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem opening stream")
+	}
+
+	return writer, nil
+}
+
+func (b *gridfsBucket) Reader(ctx context.Context, name string) (io.ReadCloser, error) {
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem resolving bucket")
+	}
+
+	writer, err := grid.OpenDownloadStreamByName(name)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem opening stream")
+	}
+
+	return writer, nil
+}
+
+func (b *gridfsBucket) Put(ctx context.Context, name string, input io.Reader) error {
+	if b.opts.DryRun {
+		return nil
+	}
+
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return errors.Wrap(err, "problem resolving bucket")
+	}
+
+	if _, err = grid.UploadFromStream(name, input); err != nil {
+		return errors.Wrap(err, "problem uploading file")
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.Reader(ctx, name)
+}
+
+func (b *gridfsBucket) Upload(ctx context.Context, name, path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return errors.Wrapf(err, "problem opening file %s", name)
+	}
+	defer f.Close()
+
+	return errors.WithStack(b.Put(ctx, name, f))
+}
+
+func (b *gridfsBucket) Download(ctx context.Context, name, path string) error {
+	reader, err := b.Reader(ctx, name)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err = os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+		return errors.Wrapf(err, "problem creating enclosing directory for '%s'", path)
+	}
+
+	f, err := os.Create(path)
+	if err != nil {
+		return errors.Wrapf(err, "problem creating file '%s'", path)
+	}
+	_, err = io.Copy(f, reader)
+	if err != nil {
+		_ = f.Close()
+		return errors.Wrap(err, "problem copying data")
+	}
+
+	return errors.WithStack(f.Close())
+}
+
+func (b *gridfsBucket) Push(ctx context.Context, local, remote string) error {
+	localPaths, err := walkLocalTree(ctx, local)
+	if err != nil {
+		return errors.Wrap(err, "problem finding local paths")
+	}
+
+	for _, path := range localPaths {
+		target := filepath.Join(remote, path)
+		_ = b.Remove(ctx, target)
+		if err = b.Upload(ctx, target, filepath.Join(local, path)); err != nil {
+			return errors.Wrapf(err, "problem uploading '%s' to '%s'", path, target)
+		}
+	}
+
+	if b.opts.DeleteOnSync && !b.opts.DryRun {
+		return errors.Wrapf(os.RemoveAll(local), "problem removing '%s' after push", local)
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) Pull(ctx context.Context, local, remote string) error {
+	iter, err := b.List(ctx, remote)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	keys := []string{}
+
+	for iter.Next(ctx) {
+		item := iter.Item()
+		name := filepath.Join(local, item.Name()[len(remote)+1:])
```

Comment:
I would use the filepath.Rel function -> `filepath.Rel(remote, item.Name())`

Reply:
I think this would lead to weird outcomes on windows